theano/sandbox/cuda/blas.py

```diff
@@ -14,6 +14,220 @@
                                            gpu_contiguous)
 from theano.tensor import as_tensor_variable
 
+class BatchedDotOp(GpuOp):
+
+    def __str__(self):
+        return 'BatchedDotOp'
+
+    def __eq__(self, other):
+        return type(self) == type(other)
+
+    def __hash__(self):
+        return hash(type(self))
+
+    def make_node(self, inp1, inp2):
+        inp1 = gpu_contiguous(as_cuda_ndarray_variable(inp1))
+        inp2 = gpu_contiguous(as_cuda_ndarray_variable(inp2))
+
+        assert inp1.dtype == "float32"
+        assert inp2.dtype == "float32"
+        assert inp1.ndim == 3 # (batch, a, b)
+        assert inp2.ndim == 3
+
+        return theano.Apply(self, [inp1, inp2], [self.output_type(inp1)()])
+
+    def output_type(self, inp):
+        return CudaNdarrayType(broadcastable=[False] * inp.type.ndim)
```

Comment:
That is ok, but not the best. If inp1 and inp2 have broadcastable dimensions, we can infer some broadcastable dims for the outputs. If inp1.broad[0] or inp2.broad[0], then we know the output should be broadcastable in that dim.

If inp1.broad[1] is broad, out.broad[1] is broad. If inp2.broad[2] is broad, out.broad[2] is broad.